addons/l10n_es_edi_facturae/models/l10n_es_edi_facturae_certificate.py

```diff
@@ -0,0 +1,117 @@
+
+    @api.model
+    def _decode_certificate(self):
+        """
+        Return certificate data
+        :return tuple: private_key, certificate
+        """
+        self.ensure_one()
+        content, password = b64decode(self.with_context(bin_size=False).content), self.password.encode()
+        private_key, certificate, *_dummy = pkcs12.load_key_and_certificates(content, password, backend=default_backend())
+        return private_key, certificate
+
+    # -------------------------------------------------------------------------
+    # LOW-LEVEL METHODS
+    # -------------------------------------------------------------------------
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        certificates = super().create(vals_list)
+        for certificate in certificates:
+            try:
+                _key, certif = certificate._decode_certificate()
+                cert_date_start, cert_date_end, serial_number = certif.not_valid_before, certif.not_valid_after, certif.serial_number
+            except ValueError:
+                raise UserError(_('There has been a problem with the certificate, some usual problems can be:\n'
+                                  '\t- The password given or the certificate are not valid.\n'
+                                  '\t- The certificate content is invalid.'))
+            if datetime.now() > cert_date_end:
+                raise UserError(_('The certificate is expired since %s') % certificate.date_end)
+            # Assign extracted values from the certificate
+            certificate.write({'serial_number': serial_number, 'date_start': cert_date_start, 'date_end': cert_date_end, })
+        return certificates
+
+    # -------------------------------------------------------------------------
+    # BUSINESS METHODS                                                        #
+    # -------------------------------------------------------------------------
+    def sign_xml(self, edi_data, signature_data):
+        """
+        Signs the given XML data with the certificate and private key.
+        :param str edi_data: The XML data to sign.
+        :param dict signature_data: The signature data to use.
+        :return str: The signed XML data as a sting.
+        """
+        self.ensure_one()
+        if not (self.date_start < fields.Datetime.now() < self.date_end):
+            raise UserError('Facturae certificate date is not valid, its validity has probably expired')
+
+        root = deepcopy(edi_data)
+        cert_private, cert_public = self._decode_certificate()
+        public_key_numbers = cert_public.public_key().public_numbers()
+        pattern = re.compile(r'(?<=\()([A-Z]{1,2}=.*)(?=\))')
+        issuer = ', '.join(sorted(match.group() for match in (pattern.search(str(element)) for element in cert_public.issuer.rdns) if match))
+
+        # Identifiers
+        document_id = f"Document-{str(uuid3(NAMESPACE_OID, str(edi_data)))}"
```

Comment:
True, f-strings call the __str__ method, as does str() but uuid3 doesn't, and the str() is inside it

Here I just want to generate a unique ID somewhat dependent on the content, not really following a specific ISO
I mostly want to avoid duplicated ids